exporter/collector/logs.go

```diff
@@ -562,54 +531,126 @@ type httpRequestLog struct {
-	req.Header.Set("User-Agent", parsedHTTPRequest.UserAgent)
 
-	httpRequest := &logging.HTTPRequest{
-		Request:                        req,
+	pb := &logtypepb.HttpRequest{
+		RequestMethod:                  parsedHTTPRequest.RequestMethod,
+		RequestUrl:                     fixUTF8(parsedHTTPRequest.RequestURL),
 		RequestSize:                    parsedHTTPRequest.RequestSize,
 		Status:                         parsedHTTPRequest.Status,
 		ResponseSize:                   parsedHTTPRequest.ResponseSize,
-		LocalIP:                        parsedHTTPRequest.ServerIP,
-		RemoteIP:                       parsedHTTPRequest.RemoteIP,
+		UserAgent:                      parsedHTTPRequest.UserAgent,
+		ServerIp:                       parsedHTTPRequest.ServerIP,
+		RemoteIp:                       parsedHTTPRequest.RemoteIP,
+		Referer:                        parsedHTTPRequest.Referer,
 		CacheHit:                       parsedHTTPRequest.CacheHit,
 		CacheValidatedWithOriginServer: parsedHTTPRequest.CacheValidatedWithOriginServer,
+		Protocol:                       "HTTP/1.1",
 		CacheFillBytes:                 parsedHTTPRequest.CacheFillBytes,
 		CacheLookup:                    parsedHTTPRequest.CacheLookup,
 	}
 	if parsedHTTPRequest.Latency != "" {
 		latency, err := time.ParseDuration(parsedHTTPRequest.Latency)
-		if err == nil {
-			httpRequest.Latency = latency
+		if err == nil && latency != 0 {
+			pb.Latency = durationpb.New(latency)
 		}
 	}
+	return pb, nil
+}
 
-	return httpRequest, nil
+// toProtoStruct converts v, which must marshal into a JSON object,
+// into a Google Struct proto.
+// Mostly copied from
+// https://github.com/googleapis/google-cloud-go/blob/69705144832c715cf23832602ad9338b911dff9a/logging/logging.go#L577
+func toProtoStruct(v interface{}) (*structpb.Struct, error) {
+	// v is a Go value that supports JSON marshaling. We want a Struct
+	// protobuf. Some day we may have a more direct way to get there, but right
+	// now the only way is to marshal the Go value to JSON, unmarshal into a
+	// map, and then build the Struct proto from the map.
+	jb, err := json.Marshal(v)
+	if err != nil {
+		return nil, fmt.Errorf("logging: json.Marshal: %w", err)
+	}
+	var m map[string]interface{}
+	err = json.Unmarshal(jb, &m)
+	if err != nil {
+		return nil, fmt.Errorf("logging: json.Unmarshal: %w", err)
+	}
+	return jsonMapToProtoStruct(m), nil
+}
+
+// Copied from https://github.com/googleapis/google-cloud-go/blob/69705144832c715cf23832602ad9338b911dff9a/logging/logging.go#L604
+func jsonMapToProtoStruct(m map[string]interface{}) *structpb.Struct {
+	fields := map[string]*structpb.Value{}
+	for k, v := range m {
+		fields[k] = jsonValueToStructValue(v)
+	}
+	return &structpb.Struct{Fields: fields}
+}
+
+// Copied from https://github.com/googleapis/google-cloud-go/blob/69705144832c715cf23832602ad9338b911dff9a/logging/logging.go#L612
+func jsonValueToStructValue(v interface{}) *structpb.Value {
```

Comment:
I'm not sure if this code predates this function, but wouldn't [`structpb.NewStruct()`](https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb#NewStruct) work instead of this?

Reply:
It looks like it works, so I changed it to use that.